### ISSUE TITLE
Inconsistent application of "Variable not allowed in type expression Pylance(reportInvalidTypeForm)"

### ISSUE THREAD

**Issue by mvanderlee** [microsoft/pyright#7268]
**Describe the bug**
New types created via `typing.NewType` can incorrectly trigger `Variable not allowed in type expression Pylance(reportInvalidTypeForm)`

in particular, if I have a function that assigns the result or `typing.NewType` to a variable, and then return that variable, it's interpreted as `Never` instead of the new type. If I return the result immediately without assigning it to a variable first, then it works correctly.

**Code or Screenshots**
![image](https://github.com/microsoft/pyright/assets/918128/421712fb-b137-4e34-89d0-11a93e74ba9e)

```python
import typing

# A documented by `typing.NewType`
UserId = typing.NewType('UserId', int)

def name_by_id(user_id: UserId) -> str:
    pass

###########################################
def create_user_id_type():
    return typing.NewType('UserId2', int)

UserId2 = create_user_id_type()

def name_by_id(user_id: UserId2) -> str:
    pass

###########################################
def create_user_id_type3():
    user_id_type =  typing.NewType('UserId3', int)

    return user_id_type

UserId3 = create_user_id_type3()

def name_by_id(user_id: UserId3) -> str:
    pass

```

**VS Code extension or command-line**
VSCode with PyLance
VSCode 1.86.1
PyLance v2024.2.2

**Comment by erictraut** [comment]
Pyright is working correctly here. If you return the result of a `NewType` call from a function, it is considered a dynamic value. It's similar to defining a class dynamically within a function and then returning it from that function.

The `NewType` call is intended to be used in the global (module-level) context. You're using it inappropriately in your code above in the `create_user_id_type3` method, and pyright is correct to flag this as an error.

**Comment by mvanderlee** [comment]
Okay so how do I disable the warning. This breaks my workflow with marshmallow-dataclass.

**Comment by erictraut** [comment]
Rather than disabling the warning, I recommend changing your code so it complies with the Python typing specification. Otherwise you may see other unexpected or undesirable behaviors at type checking time.

If you want to disable the warning, you can use [project-level configuration settings](https://microsoft.github.io/pyright/#/configuration) or [file-level or line-level comments](https://microsoft.github.io/pyright/#/comments).

**Comment by mvanderlee** [comment]
This has been working fine for multiple years now. Can you tell me where it says that your interpretation is correct? I don't see this in PEP-484?

**Comment by erictraut** [comment]
PEP 484 is a historical document. It has many areas that have been superseded and clarified subsequent to its acceptance. 

The definitive Python typing spec can be found [here](https://typing.readthedocs.io/en/latest/spec/index.html). We have been working to improve the clarity of the spec (address under-specified areas) and implement [conformance tests](https://github.com/python/typing/blob/main/conformance/README.md) for type checkers. I'm a member of the [recently-formed](https://peps.python.org/pep-0729/) five-person Typing Council, which is spearheading this effort.

If you would like to discuss aspects of the typing spec with the broader typing community, the [Python typing forum](https://discuss.python.org/c/typing/32) is a good place to get involved.

**Comment by cj81499** [comment]
This has caused a headache for me.

`AnimalId = NewType("AnimalId", pydantic.UUID4)` previously worked, but now `AnimalId` has a type of `Never`. My understanding is that this is b/c `pydantic.UUID4` is implemented as `UUID4 = Annotated[UUID, UuidVersion(4)]`, but I'm not certain my understanding is correct.

To work around this, I came up with two ideas. The first was to use `AnimalId = NewType("AnimalId", uuid.UUID)`. I'd prefer to avoid this because it means I lose out on `pydantic`'s runtime validation. The second approach is the following snippet:

```python
# HACK: We want a `NewType("AnimalId", pydantic.UUID4)`, but Pyright/Pylance doesn't support that.
# https://github.com/microsoft/pyright/issues/7268
if TYPE_CHECKING:
    class AnimalId(pydantic.UUID4):
        def __init__(self, _x: pydantic.UUID4) -> None: ...
else:
    def identity(_x):
        return _x
    AnimalId = identity
```

This seems to work at both type-checking-time and runtime, but is an awful lot to write everywhere I use `NewType` like this. Is there a better approach? Or a chance of revising the spec such that cases like this are supported? Thanks!

**Comment by mvanderlee** [comment]
@cj81499 Why can't you just use `AnimalId = pydantic.UUID4`?

**Comment by cj81499** [comment]
> @cj81499 Why can't you just use `AnimalId = pydantic.UUID4`?

In my application, we use `UUID`s to uniquely identify a variety of different things (eg: Cats, Dogs, Cars, Boats, etc). `NewType` "protects" developers from accidentally passing a `UUID` meant to identify one thing (a Cat) to a function that expects a `UUID` meant to identify something else (a Boat).

For example:
```python
from typing import NewType

CatId = NewType("CatId", int)
BoatId = NewType("BoatId", int)

def cats_only(id: CatId) -> None: ...

cats_only(CatId(1))
cats_only(BoatId(1)) # fails type checking b/c `BoatId` isn't a `CatId`
cats_only(1) # fails type checking b/c `int` isn't a `CatId`
```

If I'd instead defined `CatId` and `BoatId` as follows, the code would pass type checking.

```python
CatId = int
BoatId = int
```

try for yourself (with `mypy`): https://mypy-play.net/?mypy=latest&python=3.12&flags=strict&gist=478722cb6b96855da88594c48229b58c

**Comment by mvanderlee** [comment]
I think I understand.

Keep in mind that `Annotated` is not a real type. It's just a way to attach metadata to a type that a 3rd party can use.
So I'd suggest `CatId = NewType('CatId', uuid.UUID)`

In this case the pydantic annotation is used by pydantic to handle serialization. So you could create your own annotation per model. But if you really need want your `Id` to be the annotated type itself, then you can this: `DogId = Annotated[NewType('DogId', uuid.UUID), pydantic.UUID4]`

Full example

```python
from typing import Annotated, NewType
import uuid
import pydantic

CatId = NewType('CatId', uuid.UUID)
DogId = Annotated[NewType('DogId', uuid.UUID), pydantic.UUID4]

class Cat(pydantic.BaseModel):
    id: Annotated[CatId, pydantic.UUID4]
    name: str


class Dog(pydantic.BaseModel):
    id: DogId
    name: str
```

**Comment by cj81499** [comment]
`DogId = Annotated[NewType('DogId', uuid.UUID), pydantic.UUID4]` is exactly what I was looking for!

Thanks @mvanderlee!

**Comment by cj81499** [comment]
Update: may have spoken too soon. This doesn't actually seem to run pydantic's validation.

In the following example, "dog ta" works, but "uuid4 ta" fails w/ a validation error. I would expect both to fail.

```python
DogId = Annotated[NewType("DogId", uuid.UUID), pydantic.UUID4]


DOG_ID_TA = pydantic.TypeAdapter(DogId)
UUID4_TA = pydantic.TypeAdapter(pydantic.UUID4)

namespace = uuid.uuid4()
my_uuid = uuid.uuid3(namespace, "name")
s = str(my_uuid)
print(s)
print("dog ta", DOG_ID_TA.validate_python(s))
print("uuid4 ta", UUID4_TA.validate_python(s))
```

**Comment by mvanderlee** [comment]
Ah, makes sense. I should've followed my own advice 😅  Pydantic.UUID4 is an annoted type, which we can't use in this context.

This might work?
```python
from pydantic.types import UuidVersion
DogId = Annotated[NewType("DogId", uuid.UUID), UuidVersion(4)]
```

It doesn't work in your example though, but that's because internally pydantic does an `isinstance` on the NewType, which does not, and afaik has never worked in Python. 
So I'm assuming your provided example was just a troubleshooting step you tried, and not a reflection on the actual code.

**Comment by cj81499** [comment]
Aside from "Cat" being used as a standin for the real thing I'm identifying, `CatId = NewType("CatId", pydantic.UUID4)` is verbatim code that _used to_ work with `pyright`, but stopped working recently (still seems to work with `mypy`, but VS Code is providing awful completions as a result of `pyright`'s change).

Like you said, that code fails with some `isinstance` issue deep within `pydantic`'s internals.

I've also just realized that my original workaround does not work at runtime because I use `CatId | None`, but functions don't support the `|` operator. I think `pydantic` may also have a hard time with this, since I'm pretty sure it doesn't have any runtime indication that a `CatId` is to be treated as a `UUID4`.
